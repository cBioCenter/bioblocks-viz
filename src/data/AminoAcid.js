"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var AminoAcid = /** @class */ (function () {
    function AminoAcid(fullName, singleLetterCode, threeLetterCode, codons) {
        this.fullName = fullName;
        this.singleLetterCode = singleLetterCode;
        this.threeLetterCode = threeLetterCode;
        this.codons = codons;
    }
    // public static readonly Stop = new AminoAcid('Stop', '*', '***', ['TAA', 'TGA', 'TAG']);
    // public static readonly Gap = new AminoAcid('Gap', '-', '---', []);
    // public static readonly Unknown = new AminoAcid('Unknown', '?', '???', []);
    AminoAcid.getAllCanonicalAminoAcids = function () {
        return Object.keys(AminoAcid).reduce(function (acc, staticVarName) {
            var staticValue = AminoAcid[staticVarName];
            if (staticValue instanceof AminoAcid) {
                acc.push(staticValue);
            }
            return acc;
        }, new Array());
    };
    AminoAcid.fromSingleLetterCode = function (code) {
        return AminoAcid.getAllCanonicalAminoAcids().find(function (aa) {
            if (aa.singleLetterCode.toUpperCase() === code.toUpperCase()) {
                return true;
            }
            return false;
        });
    };
    AminoAcid.fromThreeLetterCode = function (code) {
        AminoAcid.getAllCanonicalAminoAcids().forEach(function (aa) {
            if (aa.threeLetterCode.toUpperCase() === code.toUpperCase()) {
                return aa;
            }
        });
        return undefined;
    };
    AminoAcid.fromCodon = function (codon) {
        AminoAcid.getAllCanonicalAminoAcids().forEach(function (aa) {
            aa.codons.forEach(function (codonStr) {
                if (codonStr.toUpperCase() === codon.toUpperCase()) {
                    return aa;
                }
            });
        });
        return undefined;
    };
    AminoAcid.Alanine = new AminoAcid('Alanine', 'A', 'ALA', ['GCT', 'GCC', 'GCA', 'GCG']);
    AminoAcid.Arginine = new AminoAcid('Arginine', 'R', 'ARG', ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG']);
    AminoAcid.Asparagine = new AminoAcid('Asparagine', 'N', 'ASN', ['AAT', 'AAC']);
    AminoAcid.AsparticAcid = new AminoAcid('Aspartic Acid', 'D', 'ASP', ['GAT', 'GAC']);
    AminoAcid.Cysteine = new AminoAcid('Cysteine', 'C', 'CYS', ['TGT', 'TGC']);
    AminoAcid.Glutamine = new AminoAcid('Glutamine', 'Q', 'GLN', ['CAA', 'CAG']);
    AminoAcid.GlutamicAcid = new AminoAcid('Glutamic Acid', 'E', 'GLU', ['GAA', 'GAG']);
    AminoAcid.Glycine = new AminoAcid('Glycine', 'G', 'GLY', ['GGT', 'GGC', 'GGA', 'GGG']);
    AminoAcid.Histidine = new AminoAcid('Histidine', 'H', 'HIS', ['CAT', 'CAC']);
    AminoAcid.Isoleucine = new AminoAcid('Isoleucine', 'I', 'ILE', ['ATT', 'ATC', 'ATA']);
    AminoAcid.Leucine = new AminoAcid('Leucine', 'L', 'LEU', ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG']);
    AminoAcid.Lysine = new AminoAcid('Lysine', 'K', 'LYS', ['AAA', 'AAG']);
    AminoAcid.Methionine = new AminoAcid('Methionine', 'M', 'MET', ['ATG']);
    AminoAcid.Phenylalanine = new AminoAcid('Phenylalanine', 'F', 'PHE', ['TTT', 'TTC']);
    AminoAcid.Proline = new AminoAcid('Proline', 'P', 'PRO', ['CCT', 'CCC', 'CCA', 'CCG']);
    AminoAcid.Serine = new AminoAcid('Serine', 'S', 'SER', ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC']);
    AminoAcid.Threonine = new AminoAcid('Threonine', 'T', 'THR', ['ACT', 'ACC', 'ACA', 'ACG']);
    AminoAcid.Tryptophan = new AminoAcid('Tryptophan', 'W', 'TRP', ['TGG']);
    AminoAcid.Tyrosine = new AminoAcid('Tyrosine', 'Y', 'TYR', ['TAT', 'TAC']);
    AminoAcid.Valine = new AminoAcid('Valine', 'V', 'VAL', ['GTT', 'GTC', 'GTA', 'GTG']);
    // public static readonly Unknown = new AminoAcid('Unknown', 'X', 'XXX', []);
    AminoAcid.ALL_FULL_NAMES = function () { return AminoAcid.getAllCanonicalAminoAcids().map(function (aa) { return aa.fullName; }); };
    AminoAcid.ALL_1LETTER_CODES = function () { return AminoAcid.getAllCanonicalAminoAcids().map(function (a) { return a.singleLetterCode; }); };
    AminoAcid.ALL_3LETTER_CODES = function () { return AminoAcid.getAllCanonicalAminoAcids().map(function (a) { return a.threeLetterCode; }); };
    AminoAcid.ALL_CODONS = function () { return lodash_1.flatten(AminoAcid.getAllCanonicalAminoAcids().map(function (a) { return a.codons; })); };
    return AminoAcid;
}());
exports.AminoAcid = AminoAcid;
// Gets mismatches between a PDB file and coupling scores.
// @returns Array of mismatches - empty if none are found.
exports.getPDBAndCouplingMismatch = function (pdbData, couplingScores) {
    var pdbSequence = pdbData.sequence;
    var couplingSequence = couplingScores.sequence;
    var mismatches = exports.getSequenceMismatch(pdbSequence, couplingSequence);
    if (mismatches.length === 0) {
        return pdbData.getResidueNumberingMismatches(couplingScores);
    }
    return mismatches;
};
// Given two **equal-length** Amino Acid sequences, returns all mismatches.
exports.getSequenceMismatch = function (firstSequence, secondSequence) {
    var mismatches = new Array();
    if (firstSequence.length === secondSequence.length) {
        for (var i = 0; i < firstSequence.length; ++i) {
            var couplingAminoAcid = AminoAcid.fromSingleLetterCode(firstSequence[i]);
            var pdbAminoAcid = AminoAcid.fromSingleLetterCode(secondSequence[i]);
            if (firstSequence[i] !== secondSequence[i] && couplingAminoAcid && pdbAminoAcid) {
                mismatches.push({
                    couplingAminoAcid: couplingAminoAcid,
                    pdbAminoAcid: pdbAminoAcid,
                    resno: i,
                });
            }
        }
    }
    return mismatches;
};
//# sourceMappingURL=AminoAcid.js.map