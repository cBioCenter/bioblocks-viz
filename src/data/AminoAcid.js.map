{"version":3,"file":"AminoAcid.js","sourceRoot":"","sources":["../../../src/data/AminoAcid.ts"],"names":[],"mappings":";;AAAA,iCAAiC;AAQjC;IA6EE,mBACW,QAAgB,EAChB,gBAAwB,EACxB,eAAuB,EACvB,MAAyB;QAHzB,aAAQ,GAAR,QAAQ,CAAQ;QAChB,qBAAgB,GAAhB,gBAAgB,CAAQ;QACxB,oBAAe,GAAf,eAAe,CAAQ;QACvB,WAAM,GAAN,MAAM,CAAmB;IACjC,CAAC;IArDJ,0FAA0F;IAC1F,qEAAqE;IACrE,6EAA6E;IAE/D,mCAAyB,GAAvC;QACE,OAAO,MAAM,CAAC,IAAI,CAAC,SAAS,CAAC,CAAC,MAAM,CAAC,UAAC,GAAG,EAAE,aAAa;YACtD,IAAM,WAAW,GAAG,SAAS,CAAC,aAAuC,CAAC,CAAC;YAEvE,IAAI,WAAW,YAAY,SAAS,EAAE;gBACpC,GAAG,CAAC,IAAI,CAAC,WAAW,CAAC,CAAC;aACvB;YAED,OAAO,GAAG,CAAC;QACb,CAAC,EAAE,IAAI,KAAK,EAAa,CAAC,CAAC;IAC7B,CAAC;IAEa,8BAAoB,GAAlC,UAAmC,IAAY;QAC7C,OAAO,SAAS,CAAC,yBAAyB,EAAE,CAAC,IAAI,CAAC,UAAA,EAAE;YAClD,IAAI,EAAE,CAAC,gBAAgB,CAAC,WAAW,EAAE,KAAK,IAAI,CAAC,WAAW,EAAE,EAAE;gBAC5D,OAAO,IAAI,CAAC;aACb;YAED,OAAO,KAAK,CAAC;QACf,CAAC,CAAC,CAAC;IACL,CAAC;IAEa,6BAAmB,GAAjC,UAAkC,IAAY;QAC5C,SAAS,CAAC,yBAAyB,EAAE,CAAC,OAAO,CAAC,UAAA,EAAE;YAC9C,IAAI,EAAE,CAAC,eAAe,CAAC,WAAW,EAAE,KAAK,IAAI,CAAC,WAAW,EAAE,EAAE;gBAC3D,OAAO,EAAE,CAAC;aACX;QACH,CAAC,CAAC,CAAC;QAEH,OAAO,SAAS,CAAC;IACnB,CAAC;IAEa,mBAAS,GAAvB,UAAwB,KAAa;QACnC,SAAS,CAAC,yBAAyB,EAAE,CAAC,OAAO,CAAC,UAAA,EAAE;YAC9C,EAAE,CAAC,MAAM,CAAC,OAAO,CAAC,UAAC,QAAgB;gBACjC,IAAI,QAAQ,CAAC,WAAW,EAAE,KAAK,KAAK,CAAC,WAAW,EAAE,EAAE;oBAClD,OAAO,EAAE,CAAC;iBACX;YACH,CAAC,CAAC,CAAC;QACL,CAAC,CAAC,CAAC;QAEH,OAAO,SAAS,CAAC;IACnB,CAAC;IA1EsB,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7E,kBAAQ,GAAG,IAAI,SAAS,CAAC,UAAU,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7F,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACrE,sBAAY,GAAG,IAAI,SAAS,CAAC,eAAe,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC1E,kBAAQ,GAAG,IAAI,SAAS,CAAC,UAAU,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACjE,mBAAS,GAAG,IAAI,SAAS,CAAC,WAAW,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACnE,sBAAY,GAAG,IAAI,SAAS,CAAC,eAAe,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC1E,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7E,mBAAS,GAAG,IAAI,SAAS,CAAC,WAAW,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACnE,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC5E,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC3F,gBAAM,GAAG,IAAI,SAAS,CAAC,QAAQ,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7D,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,CAAC,CAAC,CAAC;IAC9D,uBAAa,GAAG,IAAI,SAAS,CAAC,eAAe,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC3E,iBAAO,GAAG,IAAI,SAAS,CAAC,SAAS,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAC7E,gBAAM,GAAG,IAAI,SAAS,CAAC,QAAQ,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACzF,mBAAS,GAAG,IAAI,SAAS,CAAC,WAAW,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACjF,oBAAU,GAAG,IAAI,SAAS,CAAC,YAAY,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,CAAC,CAAC,CAAC;IAC9D,kBAAQ,GAAG,IAAI,SAAS,CAAC,UAAU,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IACjE,gBAAM,GAAG,IAAI,SAAS,CAAC,QAAQ,EAAE,GAAG,EAAE,KAAK,EAAE,CAAC,KAAK,EAAE,KAAK,EAAE,KAAK,EAAE,KAAK,CAAC,CAAC,CAAC;IAElG,6EAA6E;IAE/D,wBAAc,GAAG,cAAM,OAAA,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,EAAE,IAAI,OAAA,EAAE,CAAC,QAAQ,EAAX,CAAW,CAAC,EAA5D,CAA4D,CAAC;IACpF,2BAAiB,GAAG,cAAM,OAAA,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,CAAC,IAAI,OAAA,CAAC,CAAC,gBAAgB,EAAlB,CAAkB,CAAC,EAAlE,CAAkE,CAAC;IAC7F,2BAAiB,GAAG,cAAM,OAAA,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,CAAC,IAAI,OAAA,CAAC,CAAC,eAAe,EAAjB,CAAiB,CAAC,EAAjE,CAAiE,CAAC;IAC5F,oBAAU,GAAG,cAAM,OAAA,gBAAO,CAAC,SAAS,CAAC,yBAAyB,EAAE,CAAC,GAAG,CAAC,UAAA,CAAC,IAAI,OAAA,CAAC,CAAC,MAAM,EAAR,CAAQ,CAAC,CAAC,EAAjE,CAAiE,CAAC;IAwDrG,gBAAC;CAAA,AAnFD,IAmFC;AAnFY,8BAAS;AA6btB,0DAA0D;AAC1D,0DAA0D;AAC7C,QAAA,yBAAyB,GAAG,UAAC,OAAqB,EAAE,cAAiC;IAChG,IAAM,WAAW,GAAG,OAAO,CAAC,QAAQ,CAAC;IACrC,IAAM,gBAAgB,GAAG,cAAc,CAAC,QAAQ,CAAC;IACjD,IAAM,UAAU,GAAG,2BAAmB,CAAC,WAAW,EAAE,gBAAgB,CAAC,CAAC;IAEtE,IAAI,UAAU,CAAC,MAAM,KAAK,CAAC,EAAE;QAC3B,OAAO,OAAO,CAAC,6BAA6B,CAAC,cAAc,CAAC,CAAC;KAC9D;IAED,OAAO,UAAU,CAAC;AACpB,CAAC,CAAC;AAEF,2EAA2E;AAC9D,QAAA,mBAAmB,GAAG,UAAC,aAAqB,EAAE,cAAsB;IAC/E,IAAM,UAAU,GAAG,IAAI,KAAK,EAA0B,CAAC;IACvD,IAAI,aAAa,CAAC,MAAM,KAAK,cAAc,CAAC,MAAM,EAAE;QAClD,KAAK,IAAI,CAAC,GAAG,CAAC,EAAE,CAAC,GAAG,aAAa,CAAC,MAAM,EAAE,EAAE,CAAC,EAAE;YAC7C,IAAM,iBAAiB,GAAG,SAAS,CAAC,oBAAoB,CAAC,aAAa,CAAC,CAAC,CAAC,CAAC,CAAC;YAC3E,IAAM,YAAY,GAAG,SAAS,CAAC,oBAAoB,CAAC,cAAc,CAAC,CAAC,CAAC,CAAC,CAAC;YACvE,IAAI,aAAa,CAAC,CAAC,CAAC,KAAK,cAAc,CAAC,CAAC,CAAC,IAAI,iBAAiB,IAAI,YAAY,EAAE;gBAC/E,UAAU,CAAC,IAAI,CAAC;oBACd,iBAAiB,mBAAA;oBACjB,YAAY,cAAA;oBACZ,KAAK,EAAE,CAAC;iBACT,CAAC,CAAC;aACJ;SACF;KACF;IAED,OAAO,UAAU,CAAC;AACpB,CAAC,CAAC"}