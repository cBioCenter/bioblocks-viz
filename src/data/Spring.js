"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//# sourceMappingURL=Spring.js.map