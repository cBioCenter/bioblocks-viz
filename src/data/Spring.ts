export interface ISpringCategoricalColorDataInput {
  [k: string]: {
    label_colors: { [key: string]: string | number };
    label_list: string[];
  };
}

export interface ISpringCategoricalColorData {
  label_colors: { [key: string]: number };
  label_list: string[];
}

export interface ISpringGraphData {
  nodes: ISpringNode[];
  links: ISpringLink[];
}

export interface ISpringLink {
  distance: number;
  source: ISpringNode | number | string;
  target: ISpringNode | number | string;
}

export interface ISpringNode {
  category: string;
  colorHex: number;
  fixed: boolean;
  name: string;
  number: number;
  vx: number;
  vy: number;
  x: number;
  y: number;
}
