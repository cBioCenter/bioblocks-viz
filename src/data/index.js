"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./AminoAcid"), exports);
tslib_1.__exportStar(require("./Nucleotide"), exports);
tslib_1.__exportStar(require("./AnatomogramMapping"), exports);
tslib_1.__exportStar(require("./Bioblocks1DSection"), exports);
tslib_1.__exportStar(require("./BioblocksConfig"), exports);
tslib_1.__exportStar(require("./BioblocksData"), exports);
tslib_1.__exportStar(require("./BioblocksPDB"), exports);
tslib_1.__exportStar(require("./CodonTable"), exports);
tslib_1.__exportStar(require("./CouplingContainer"), exports);
tslib_1.__exportStar(require("./Dataset"), exports);
tslib_1.__exportStar(require("./IUPACData"), exports);
tslib_1.__exportStar(require("./Markers"), exports);
tslib_1.__exportStar(require("./PfamProtein"), exports);
tslib_1.__exportStar(require("./PlotlyData"), exports);
tslib_1.__exportStar(require("./Protein"), exports);
tslib_1.__exportStar(require("./Seq"), exports);
tslib_1.__exportStar(require("./SeqIO"), exports);
tslib_1.__exportStar(require("./SeqRecord"), exports);
tslib_1.__exportStar(require("./TintedBioblocks1DSection"), exports);
tslib_1.__exportStar(require("./event"), exports);
//# sourceMappingURL=index.js.map