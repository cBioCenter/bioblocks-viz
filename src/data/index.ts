export * from './AminoAcid';
export * from './Chell1DSection';
export * from './ChellConfig';
export * from './ChellData';
export * from './ChellPDB';
export * from './CouplingContainer';
export * from './PlotlyData';
export * from './Protein';
export * from './Spring';
export * from './TintedChell1DSection';

export * from './event';
