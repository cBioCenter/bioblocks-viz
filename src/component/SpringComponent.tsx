import * as d3 from 'd3';
import * as PIXI from 'pixi.js';
import * as React from 'react';
import CellContext, { initialCellContext } from '../context/CellContext';
import { SPRING_DATA_TYPE } from '../data/chell-data';
import { ISpringGraphData, ISpringLink, ISpringNode } from '../data/Spring';
import { withDefaultProps } from '../helper/ReactHelper';

export const defaultSpringProps = {
  canvasBackgroundColor: 0xcccccc,
  data: new Object({
    links: new Array<ISpringLink>(),
    nodes: new Array<ISpringNode>(),
  }) as SPRING_DATA_TYPE,
  height: 450,
  ...initialCellContext,
  padding: 0,
  selectedCategory: '',
  width: 450,
};

export type SpringComponentProps = {} & typeof defaultSpringProps;

export const SpringComponentWithDefaultProps = withDefaultProps(
  defaultSpringProps,
  class SpringComponentClass extends React.Component<SpringComponentProps, any> {
    protected pixiApp: PIXI.Application = new PIXI.Application();

    protected canvasElement?: HTMLCanvasElement;

    protected nodeSprites: PIXI.Container = new PIXI.Container();
    protected edgeSprites: PIXI.Container = new PIXI.Container();

    constructor(props: SpringComponentProps) {
      super(props);
      this.state = {
        ...this.state,
      };
    }

    public componentDidMount() {
      const { height, width } = this.props;
      this.pixiApp = new PIXI.Application(width, height, {
        backgroundColor: this.props.canvasBackgroundColor,
        view: this.canvasElement,
      });

      const { data, selectedCategory } = this.props;
      if (data) {
        this.pixiApp.stage.removeChildren();

        this.nodeSprites = new PIXI.Container();
        this.edgeSprites = new PIXI.Container();

        this.generateNodeSprites(data.nodes, this.nodeSprites, selectedCategory);
        this.generateLinesSprite(data.links, this.edgeSprites, selectedCategory);

        this.centerCanvas(data);

        this.pixiApp.stage.addChild(this.edgeSprites);
        this.pixiApp.stage.addChild(this.nodeSprites);
      }
    }

    public componentDidUpdate(prevProps: SpringComponentProps, prevState: any) {
      const { data, selectedCategory } = this.props;
      const isNewData = data && data !== prevProps.data;
      if (isNewData) {
        this.pixiApp.stage.removeChildren();

        this.nodeSprites = new PIXI.Container();
        this.edgeSprites = new PIXI.Container();

        this.generateNodeSprites(data.nodes, this.nodeSprites, selectedCategory);
        this.generateLinesSprite(data.links, this.edgeSprites, selectedCategory);

        this.centerCanvas(data);

        this.pixiApp.stage.addChild(this.edgeSprites);
        this.pixiApp.stage.addChild(this.nodeSprites);
      } else if (selectedCategory !== prevProps.selectedCategory) {
        this.updateNodeSprites(data.nodes, this.nodeSprites, (node: ISpringNode) => node.category === selectedCategory);
        this.edgeSprites.removeChildren();
        this.generateLinesSprite(data.links, this.edgeSprites, selectedCategory);
        this.centerCanvas(data);
      } else if (this.props.currentCells !== prevProps.currentCells) {
        this.updateNodeSprites(
          data.nodes,
          this.nodeSprites,
          (node: ISpringNode) => this.props.currentCells.indexOf(node.number) !== -1,
        );
        this.edgeSprites.removeChildren();
        this.generateLinesSprite(data.links, this.edgeSprites, selectedCategory);
        this.centerCanvas(data);
      }
    }

    public render() {
      const { height, padding, width } = this.props;
      return (
        <div id="SpringComponent" style={{ padding }}>
          {<canvas ref={el => (this.canvasElement = el ? el : undefined)} style={{ height, width }} />}
        </div>
      );
    }

    protected generateLinesSprite(links: ISpringLink[] = [], container: PIXI.Container, category?: string) {
      const lines = new PIXI.Graphics();
      const { height, width } = this.props;
      for (const link of links) {
        const source = link.source as ISpringNode;
        const target = link.target as ISpringNode;

        if (category && source.category !== category && target.category !== category) {
          continue;
        }
        lines.lineStyle(5, 0xff0000, 1);
        lines.moveTo(source.x, source.y);
        lines.lineTo(target.x, target.y);
      }
      const linesBounds = lines.getBounds();
      const textureRect = new PIXI.Rectangle(
        linesBounds.x,
        linesBounds.y,
        Math.max(width, linesBounds.width),
        Math.max(height, linesBounds.height),
      );
      const linesTexture = this.pixiApp.renderer.generateTexture(
        lines,
        PIXI.SCALE_MODES.LINEAR,
        width / height,
        textureRect,
      );
      const linesSprite = new PIXI.Sprite(linesTexture);
      linesSprite.x = textureRect.x;
      linesSprite.y = textureRect.y;
      container.addChild(linesSprite);
    }

    protected generateNodeSprites(nodes: ISpringNode[] = [], container: PIXI.Container, category?: string) {
      const SPRITE_IMG_SIZE = 32;
      const scaleFactor = (0.5 * 32) / SPRITE_IMG_SIZE;

      // TODO: Evaluate ParticleContainer is PIXI v5. The v4 version doesn't play nice with sprites rendered via PIXI.Graphics.
      // this.sprites = new PIXI.particles.ParticleContainer(data.nodes.length);

      for (const node of nodes) {
        const nodeTexture = new PIXI.Graphics();
        nodeTexture.beginFill(node.colorHex);
        nodeTexture.drawCircle(0, 0, SPRITE_IMG_SIZE / 2);
        nodeTexture.endFill();
        const sprite = new PIXI.Sprite(this.pixiApp.renderer.generateTexture(nodeTexture));
        sprite.x = node.x;
        sprite.y = node.y;
        if (category && node.category !== category) {
          sprite.alpha = 0.1;
        }
        sprite.anchor.set(0.5, 0.5);
        sprite.interactive = true;
        sprite.scale.set(scaleFactor);
        container.addChild(sprite);
      }
    }

    protected updateNodeSprites(
      nodes: ISpringNode[] = [],
      container: PIXI.Container,
      conditionFn: (node: ISpringNode) => boolean,
    ) {
      for (let i = 0; i < container.children.length; ++i) {
        const node = nodes[i];
        const sprite = container.children[i];
        sprite.alpha = conditionFn(node) ? 1 : 0.1;
      }
    }

    protected centerCanvas(data: ISpringGraphData) {
      const { height, width } = this.props;
      const allXs = data.nodes.map(node => node.x);
      const allYs = data.nodes.map(node => node.y);

      const max = {
        x: d3.max(allXs) as number,
        y: d3.max(allYs) as number,
      };
      const min = {
        x: d3.min(allXs) as number,
        y: d3.min(allYs) as number,
      };

      const dx = max.x - min.x + 50;
      const dy = max.y - min.y + 50;

      const scale = 0.85 / Math.max(dx / width, dy / height);

      const delta = {
        scale: scale - this.nodeSprites.scale.x,
        x: width / 2 - ((max.x + min.x) / 2) * scale - this.nodeSprites.position.x,
        y: height / 2 + 30 - ((max.y + min.y) / 2) * scale - this.nodeSprites.position.y,
      };

      this.nodeSprites.position.x += delta.x;
      this.nodeSprites.position.y += delta.y;
      this.nodeSprites.scale.x += delta.scale;
      this.nodeSprites.scale.y += delta.scale;
      this.edgeSprites.position = this.nodeSprites.position;
      this.edgeSprites.scale = this.nodeSprites.scale;
    }
  },
);

// TODO The required props should be discernable from `withDefaultProps` without needing to duplicate.
// However the Context consumer syntax is still new to me and I can't find the right combination :(
type requiredProps = Partial<typeof defaultSpringProps> &
  Required<Omit<SpringComponentProps, keyof typeof defaultSpringProps>>;

const SpringComponent = (props: requiredProps) => (
  <CellContext.Consumer>
    {({ currentCells }) => <SpringComponentWithDefaultProps {...props} currentCells={currentCells} />}
  </CellContext.Consumer>
);

export default SpringComponent;
export { SpringComponent };
