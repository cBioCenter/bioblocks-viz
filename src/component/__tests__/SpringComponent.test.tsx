import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';

import { ISpringGraphData } from '../../data/Spring';
import SpringComponent from '../SpringComponent';

describe('SpringComponent', () => {
  it('Should match existing snapshot when given simple data.', () => {
    const data: ISpringGraphData = { nodes: [], links: [] };
    expect(toJson(shallow(<SpringComponent data={data} />))).toMatchSnapshot();
  });
});
