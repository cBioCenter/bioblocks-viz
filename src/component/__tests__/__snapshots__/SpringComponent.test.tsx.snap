// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpringComponent Should match existing snapshot when given simple data. 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
